compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java

```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
         } catch (IOException e) {
             throw new RuntimeException("Error while accessing Distribution cache: " + e.getMessage());
         }
+
+        List<Path> incompatibleVersions = new ArrayList<>();
+
+        if (!versions.isEmpty()) {
+            for (Path ver : versions) {
+                Path pkgJsonPath = getPackagePath(org.value(), name.value(), ver.getFileName().toString())
+                        .resolve(ProjectConstants.PACKAGE_JSON);
+
+                if (Files.exists(pkgJsonPath)) {
+                    String packageVer = BalaFiles.readPkgJson(pkgJsonPath).getBallerinaVersion();
+                    String packVer = RepoUtils.getBallerinaShortVersion();
+
+                    if (!packageVer.equals(packVer)) {
+                        if (!isCompatible(packageVer, packVer)) {
+                            incompatibleVersions.add(ver);
+                        }
+                    }
+                } else {
+                    incompatibleVersions.add(ver);
+                }
+            }
+        }
+
+        versions.removeAll(incompatibleVersions);
         return pathToVersions(versions);
     }
 
+    private boolean isCompatible(String packageVer, String packVer) {
+        if (!packageVer.equals(packVer)) {
+            String packageStage = packageVer.substring(0, packageVer.length() - 1);
+            String packStage = packVer.substring(0, packVer.length() - 1);
+
+            // If the stages are equal, we need to check the versions
+            if (packageStage.equals(packStage)) {
+                String packageVerValue = packageVer.substring(packageVer.length() - 1);
```

Comment:
shall we rename the variables to be more explanatory? e.g. `pkgPlatformVer` and `distPlatformVer`